thewalrus/internal_modes/distinguishable_pnr_prob.py

```diff
@@ -0,0 +1,254 @@
+# Copyright 2021-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Function for calculating probability of distinguishable squeezing Gaussian Boson sampling experiments
+"""
+import numpy as np
+import numba
+
+from ..symplectic import passive_transformation, squeezing
+from ..quantum import Qmat
+from .._hafnian import nb_binom
+from .useful_tools import nb_Qmat, fact
+
+
+@numba.jit(nopython=True, cache=True)
+def f_all(E, n):
+    """
+    evaluate the polyonial coefficients of the function in the eigevalue-trace formula
+    Args:
+        E (array): eigevavlues of AX
+        n (int): number of polynomial coefficients to compute
+    Returns:
+        array: polynomial coefficients
+    """
+    E_k = E.copy()
+    # Compute combinations in O(n^2log n) time
+    # code translated from thewalrus matlab script
+    count = 0
+    comb = np.zeros((2, n // 2 + 1), dtype=np.complex128)
+    comb[0, 0] = 1
+    for i in range(1, n // 2 + 1):
+        factor = E_k.sum() / (2 * i)
+        E_k *= E
+        powfactor = 1
+        count = 1 - count
+        comb[count, :] = comb[1 - count, :]
+        for j in range(1, n // (2 * i) + 1):
+            powfactor *= factor / j
+            for k in range(i * j + 1, n // 2 + 2):
+                comb[count, k - 1] += comb[1 - count, k - i * j - 1] * powfactor
+    return comb[count, :]
+
+
+@numba.jit(nopython=True)
+def guan_code(n):
+    """
+    generator for a Guan code sequence
+
+    Args:
+        n (array[int]): maximum value for each digit in the Guan code
+    Yields:
+        j_new (int): index of the changed digit
+        j_pm (int): +1 / -1 depending on whether the index increases or decreases
+    """
+    n = n[::-1]  # maximum for each digit
+    K = len(n)
+    g = np.zeros(K + 1, dtype=np.uint8)  # the guan code
+    u = np.ones(K + 1, dtype=np.int8)
+    while g[K] == 0:
+        # enumerate next gray code
+        i = 0
+        k = g[0] + u[0]
+        while k > n[i] or k < 0:
+            u[i] = -u[i]
+            i += 1
+            k = g[i] + u[i]
+            if i >= K:
+                break
+        g[i] = k
+
+        if i < K:
+            yield K - i - 1, u[i], g[-2::-1]
+
+
+# @numba.jit(nopython=True, cache=True)
```

Comment:
```suggestion
@numba.jit(nopython=True, cache=True)
```
Presumably we want this in?

Reply:
@nquesada @sduquemesa was it removed for the codecov test?

Reply:
I will look into this one!